AFQ/segmentation.py

```diff
@@ -805,17 +822,90 @@ def segment_reco(self, tg=None):
         tg = self._read_tg(tg=tg)
         fiber_groups = {}
 
-        self.move_streamlines(tg, self.reg_algo)
         # We generate our instance of RB with the moved streamlines:
         self.logger.info("Extracting Bundles")
-        rb = RecoBundles(self.moved_sl, verbose=False, rng=self.rng)
+        # If doing a presegmentation based on ROIs then initialize
+        # that segmentation and segment using ROIs, else
+        # RecoBundles based on the whole brain tractogram
+        if self.presegment_bundle_dict is not None:
+            roiseg = Segmentation(**self.presegment_kawrgs)
+            roiseg.segment(
+                    self.presegment_bundle_dict,
+                    self.tg,
+                    self.fdata,
+                    self.fbval,
+                    self.fbvec,
+                    reg_template=self.reg_template,
+                    mapping=self.mapping,
+                    reg_prealign=self.reg_prealign)
+            roiseg_fg = roiseg.fiber_groups
+        else:
+            self.move_streamlines(tg, self.reg_algo)
+            rb = RecoBundles(self.moved_sl, verbose=False, rng=self.rng)
         # Next we'll iterate over bundles, registering each one:
         bundle_list = list(self.bundle_dict.keys())
         bundle_list.remove('whole_brain')
 
         self.logger.info("Assigning Streamlines to Bundles")
         for bundle in bundle_list:
+            self.logger.info(f"Finding streamlines for {bundle}")
             model_sl = self.bundle_dict[bundle]['sl']
+
+            # If doing a presegmentation based on ROIs then initialize rb after
+            # Filtering the whole brain tractogram to pass through ROIs
+            if self.presegment_bundle_dict is not None:
+                afq_bundle_name = afd.BUNDLE_RECO_2_AFQ.get(bundle, bundle)
```

Comment:
IIUC, this makes this rather dependent on the current relationships between waypoint and recobundle bundle names, right? What would happen if we had a `bundle_dict` with waypoint ROIs with names of RB bundles not currently defined as waypoint ROIs?

Reply:
The second argument in afd.BUNDLE_RECO_2_AFQ.get(bundle, **bundle**) means that if bundle is not in BUNDLE_RECO_2_AFQ, then **bundle** is returned. So this would still work!